src/ripple/basics/XRPAmount.h

```diff
@@ -0,0 +1,298 @@
+    Copyright (c) 2012, 2013 Ripple Labs Inc.
+
+    Permission to use, copy, modify, and/or distribute this software for any
+    purpose  with  or without fee is hereby granted, provided that the above
+    copyright notice and this permission notice appear in all copies.
+
+    THE  SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+    WITH  REGARD  TO  THIS  SOFTWARE  INCLUDING  ALL  IMPLIED  WARRANTIES  OF
+    MERCHANTABILITY  AND  FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+    ANY  SPECIAL ,  DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+    WHATSOEVER  RESULTING  FROM  LOSS  OF USE, DATA OR PROFITS, WHETHER IN AN
+    ACTION  OF  CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+    OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+*/
+//==============================================================================
+
+#ifndef RIPPLE_BASICS_XRPAMOUNT_H_INCLUDED
+#define RIPPLE_BASICS_XRPAMOUNT_H_INCLUDED
+
+#include <ripple/basics/contract.h>
+#include <ripple/beast/cxx17/type_traits.h>
+#include <ripple/beast/utility/Zero.h>
+#include <ripple/json/json_value.h>
+#include <boost/operators.hpp>
+#include <boost/multiprecision/cpp_int.hpp>
+#include <cstdint>
+#include <string>
+
+namespace ripple {
+
+class XRPAmount
+    : private boost::totally_ordered <XRPAmount>
+    , private boost::additive <XRPAmount>
+    , private boost::dividable <XRPAmount, std::int64_t>
+    , private boost::modable <XRPAmount, std::int64_t>
+    , private boost::unit_steppable <XRPAmount>
+{
+public:
+    using unit_type = XRPAmount;
+    using value_type = std::int64_t;
+private:
+    value_type drops_;
+
+protected:
+    template<class Other>
+    static constexpr bool is_compatible_v = std::is_integral_v<Other>;
+
+    template<class Other>
+    using enable_if_compatible_t =
+        typename std::enable_if_t<is_compatible_v<Other>>;
+
+public:
+    XRPAmount () = default;
+    constexpr XRPAmount (XRPAmount const& other) = default;
+    constexpr XRPAmount& operator= (XRPAmount const& other) = default;
+
+    constexpr
+    XRPAmount (beast::Zero)
+        : drops_ (0)
+    {
+        static_assert(std::is_convertible_v<value_type, std::uint64_t>,
+            "conversion");
+    }
+
+    constexpr
+    XRPAmount&
+    operator= (beast::Zero)
+    {
+        drops_ = 0;
+        return *this;
+    }
+
+    constexpr
+    XRPAmount (value_type drops)
+        : drops_ (drops)
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    constexpr
+    XRPAmount (Other drops)
+        : drops_ (static_cast<value_type> (drops))
+    {
+    }
+
+    template <class Other,
+        class = enable_if_compatible_t <Other>>
+    XRPAmount&
+    operator= (Other drops)
+    {
+        drops_ = static_cast<value_type> (drops);
+        return *this;
+    }
+
+    constexpr
+    XRPAmount
+    operator*(value_type const& rhs) const
```

Comment:
This adds `XRPAmount * scalar`, we should add `scalar * XRPAmount` as well. Also, if we can multiply we should probably be able to divide as well (I'm usually against adding functions we don't use, but it does seem incomplete without it).

Reply:
The `scalar * XRPAmount` operator was defined further down, but I moved it to be a friend inside the class, so that's clearer. Division is provided by `boost::dividable`. (I would have used `boost::multiplicative`, but that doesn't give me `constexpr` multiplication.)

Reply:
I did miss the one lower down. Thanks for moving it up. Nit: Is there a reason that one is defined as a member function and one is defined as a friend? I'd define them both as friends (tho no change needed, it's a minor nit).

Reply:
There are two reasons one is a member function, and one is a friend:
1. That's how it was originally written, and I'm lazy.
2. Because I can't make them both member functions. Seriously, if I could, I would. I have a significant preference for member functions, but it's just a preference.

Reply:
I wrote a different reply first, but realized that I remembered incorrectly how I got there. The main reason I did one as a member and one as a free function, later a friend, is just that I have a preference for using member functions instead of free functions. I'm not going to change it, but I don't have any objection either.

Reply:
I'm fine with it. I'm marking this resolved.